pkg/sdkserver/sdkserver.go

```diff
@@ -69,6 +71,23 @@ var (
 	_ beta.SDKServer  = &SDKServer{}
 )
 
+type counterUpdateRequest struct {
+	// Tracks CountIncrement and / or CountDecrement Requests
+	diffList []int64
+	// Counter (as retreived during first request)
+	counter agonesv1.CounterStatus
+	// Tracks if there have been any CountSet requests
+	countSet bool
+	// Tracks if there have been any CapacitySet requests
+	capacitySet bool
+	// Capacity of the Counter (as retreived during first request) or overwritten by capacitySet.
+	capacity int64
+	// Count of the Counter (as retreived during first request) or overwritten by countSet.
+	count int64
+	// Tracks the sum of CountIncrement, CountDecrement, and/or CountSet requests from the client SDK.
+	diff int64
```

Comment:
```suggestion
	diff int64
	// which generation from https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1#ObjectMeta where this is applied.
	Generation *int64
```

If we track which generation (it's only ever increases in value), then we can know if the current version we get back from `s.gameServer()` has the applied new batch in it's state or not, so we know whether to apply the changes stored in this batch (and if we should apply it again).

Other option to be inline in above, could also add a `batchApplied: bool` (or similar name).

Reply:
Right now I'm using `counter agonesv1.CounterStatus` as the tracker of the Counter state. The Generation will change for any change to the Gameserver, regardless if it applies to a particular Counter we want to update. If we track the original Counter we can do a cmp.Equal to see if there any relevant changes to a Counter even if the Generation has changed.

Reply:
So the race condition we currently have, that I want to mitigate is:

1. We have a batch, and then we apply it.
2. We clear `gsCounterUpdates`
3. s.gameServer() is still returning the old version (it hasn't yet come back up to the Informer cache in the sdkserver).
4. Someone does a `GetCounter()`, which will return the old values not including the pending batch.

As far as I'm aware, the Informer cache doesn't immediately update if it sees a local update (although I could be wrong, but I'd be surprised) -- so I wanted to make sure we had something that would handle that period of time where it could be in flux - that's why I was suggesting https://github.com/googleforgames/agones/pull/3322/files#r1302253933 with something like `gsProcessedCounterUpdates`.

Reply:
Using gsCopy to track Generation changes (see GetCounter conversation).